torch/nn/utils/clip_grad.py

```diff
@@ -24,14 +24,17 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2):
     parameters = list(filter(lambda p: p.grad is not None, parameters))
     max_norm = float(max_norm)
     norm_type = float(norm_type)
+    if len(parameters) == 0:
+        return torch.tensor(0.)
+    device = parameters[0].grad.device
```

Comment:
Can I suggest some optimizations?
This API can be called with 100s of parameters, so performance is important IMHO.
Also, the case with parameters on different devices is likely not going to be the most common.
If param[0] is one device 0, and param[1:100] are on device 1, in the mul_() loop below we would 99 to().
Also in the loops below we would be doing empty/no-op to() calls. I am not sure what the cost is, but I think this change needs to be benchmarked.

If we keep instead a dictionary device->total_norm and in the loops below we reduce the per-device total_norm.
At the end we will find an N-device sized dictionary, which we use to get the global total_norma, and do only N to() calls to assign the per-device dictionary with the per device total_norm.
Then the mul_() loop below would just lookup the dictionary with device key.

This would remove any to() calls in the most common case (single device), and reduce the total number even in other cases.

@ailzhang for double checking 😄

Reply:
@dlibenzi good point about possibly multiple cross-device calls in some unfortunate situations, it might be good to look at later. My first approach avoided cross-device calls by using python scalar, but it had a drawback of being synchronizing even for a single device. Having a lazily-initialized dictionary adds code complexity which may not be warranted if the problem we are trying to solve is niche. This PR aims to unbreak people using model parallel, who now cannot use clip_grad_norm at all, and it is also not slower for them then the previous working version, that synchronized for every parameter. 
I've benchmarked, and there is indeed a small overhead for unconditional `.to(device)` call, maybe 3%, it's pretty noisy. Again, I don't think it's worth to get rid of it at the expense of additional code complexity.

Reply:
@ailzhang Created a PR to undo these changes on our side, so we are good for now.